wip wip wip TC trigger

andreimatei opened the pull request.

andreimatei committed:
wip panic on use after finish

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72448)
<!-- Reviewable:end -->

andreimatei committed:
logictest: improve progress output

Logictests would print some progress messages to stdout as they run.
Each file prints when it ends and, while it's running, it prints
messages when statements finish successfully (but only up to once per
2s). This patch adds a second-level timestamp to those messages, in
order to provide more clues in the output of timed out logictest
packages. Otherwise, I'm always left guessing if the package had been
making constant progress until the timeout, or if it got stuck at some
point.

Release note: None

andreimatei committed:
util/tracing: improve a benchmark

BenchmarkSpanCreation was single-threaded, measuring span creation
throughput on a single-core. Because we have the trace registry, which
acts as a central resource, the scalability of the Tracer is important
too. This patch makes the benchmark parallel.

Release note: None

andreimatei committed:
tracing: improve Tracer scalability by introducing a scalable rand

Each span gets a random span id, and each trace gets a random trace id.
rand.Int63() doesn't scale - it's a mutex around a source. This patch
switches to using the go runtime's fastrand.

The difference is noticeable. I ran the benchmarks with GOGC=off for
more useful numbers because, for this small bench, we're mostly
measuring dynamic memory allocation and GC (particularly since go 1.17
runs benchmarks with a tiny initial heap, so the GC is very frequent).

GOGC=off make bench PKG=./pkg/util/tracing BENCHES=BenchmarkSpanCreation TESTFLAGS="-cpu=1,2,4,8,16,32"

SpanCreation       2.06µs ±25%    1.99µs ±26%     ~     (p=0.310 n=5+5)
SpanCreation-2     1.67µs ±24%    1.30µs ±13%  -22.26%  (p=0.008 n=5+5)
SpanCreation-4     1.94µs ± 5%    1.46µs ± 3%  -24.49%  (p=0.008 n=5+5)
SpanCreation-8     2.12µs ± 4%    1.83µs ± 4%  -13.80%  (p=0.008 n=5+5)
SpanCreation-16    1.94µs ±18%    2.14µs ± 6%     ~     (p=0.222 n=5+5)
SpanCreation-32    2.31µs ± 2%    2.33µs ± 4%     ~     (p=0.690 n=5+5)

andreimatei committed:
colexec: use fastrand for hashing

The respective hash function had been copied from the runtime, with a
comment that we can't use the runtime's fastrand. We now have fastrand,
so let's use it for speedup and scalability.

Release note: None

andreimatei committed:
tracing: improve benchmark

Benchmark the creation of both regular children and detached children.
Their scalability is currently different, because the detached children
are added to the registry.

Release note: None

andreimatei committed:
tracing: expand the active spans registry

Before this patch, the registry only contained (local) roots. A child
span was not part of the registry directly; it was part of it indirectly
if the parent had been recording at the moment when the child was
opened.

I believe the fact that a child of a non-recording span is not part of
the registry is not principled, just path dependency. This patch makes
child spans indirectly part of the registry even if the parent is not
recording at the moment when the child is created, by having children
always register themselves with the parent, even if the parent is not
recording.

This patch also improves the scalability of creating child spans by
avoiding to attempt to remove them from the registry on Finish(). This
attempt was useless because we knew that the child span was not part of
the registry. It was also expensive, since it meant locking (and thus
synchronizing on) the registry. Thus, we now only attempt to remove
roots from the registry.

The scalability increase can be seen in the following:

GOGC=off make bench PKG=./pkg/util/tracing BENCHES=BenchmarkSpanCreation TESTFLAGS="-cpu=1,2,4,8,16,32 -count=5"

SpanCreation/detached-child=false     1.95µs ±28%  2.04µs ±24%     ~     (p=0.278 n=5+5)
SpanCreation/detached-child=false-2   1.32µs ±11%  1.23µs ±15%     ~     (p=0.310 n=5+5)
SpanCreation/detached-child=false-4   1.51µs ± 5%  0.81µs ±11%  -46.49%  (p=0.008 n=5+5)
SpanCreation/detached-child=false-8   1.86µs ± 2%  0.89µs ± 4%  -52.24%  (p=0.008 n=5+5)
SpanCreation/detached-child=false-16  2.08µs ± 5%  0.85µs ± 4%  -59.30%  (p=0.008 n=5+5)
SpanCreation/detached-child=false-32  2.21µs ± 2%  0.85µs ± 2%  -61.71%  (p=0.008 n=5+5)

Release note: None

andreimatei committed:
jobs: remove picky test

This test checks that a job record is not updated with a trace id if
tracing the job is not enabled. This test fails if tracing is generally
enabled across the cluster. The code under test is this:
https://github.com/cockroachdb/cockroach/blob/96a125ade349979f7ff7a42e23f39aa544c5529f/pkg/jobs/adopt.go#L401-L408

And, I'm trying to make it such that tracing is always enabled across
the cluster. So, I'm deleting this test, and also morally accepting that
there will usually be an extra update to the job record. Hopefully,
always getting that trace id in the job is a good thing.

Release note: None

andreimatei committed:
tracing: make a test robust against tracing being on

One logictest did not like it when tracing was on by default. This patch
makes the test configure the cluster in which it's running to have
tracing off by default.

Release note: None

andreimatei committed:
execstats: improve a test

This test was collecting a recording after finishing the span, which
won't be allowed any more. Also the span was not Finish()ed, in case of
test failure.

Release note: None
